p2p/ipld/write.go

```diff
@@ -42,12 +52,113 @@ func PutBlock(ctx context.Context, adder ipld.NodeAdder, block *types.Block) err
 	if err != nil {
 		return fmt.Errorf("failure to recompute the extended data square: %w", err)
 	}
+	// get row and col roots to be provided
+	// this also triggers adding data to DAG
+	prov := newProvider(ctx, croute, int32(squareSize*4), logger.With("height", block.Height))
+	for _, root := range eds.RowRoots() {
+		prov.Provide(plugin.MustCidFromNamespacedSha256(root))
+	}
+	for _, root := range eds.ColumnRoots() {
+		prov.Provide(plugin.MustCidFromNamespacedSha256(root))
+	}
+	// commit the batch to ipfs
+	err = batchAdder.Commit()
+	if err != nil {
+		return err
+	}
+	// wait until we provided all the roots if requested
+	<-prov.Done()
+	return prov.Err()
+}
 
-	// thanks to the batchAdder.Visit func we added to the nmt wrapper,
-	// generating the roots will start adding the data to IPFS
-	eds.RowRoots()
-	eds.ColumnRoots()
+var provideWorkers = 32
 
-	// commit the batch to ipfs
-	return batchAdder.Commit()
+type provider struct {
+	ctx  context.Context
+	done chan struct{}
+
+	err   error
+	errLk sync.RWMutex
+
+	jobs  chan cid.Cid
+	total int32
+
+	croute routing.ContentRouting
+	log    log.Logger
+	took   time.Time
+}
+
+func newProvider(ctx context.Context, croute routing.ContentRouting, toProvide int32, logger log.Logger) *provider {
+	p := &provider{
+		ctx:    ctx,
+		done:   make(chan struct{}),
+		jobs:   make(chan cid.Cid, provideWorkers),
+		total:  toProvide,
+		croute: croute,
+		log:    logger,
+	}
+	for range make([]bool, provideWorkers) {
+		go p.worker()
+	}
+	logger.Info("Started Providing to DHT")
+	p.took = time.Now()
+	return p
+}
+
+func (p *provider) Provide(id cid.Cid) {
+	select {
+	case p.jobs <- id:
+	case <-p.ctx.Done():
+	}
+}
+
+func (p *provider) Done() <-chan struct{} {
+	return p.done
+}
+
+func (p *provider) Err() error {
+	p.errLk.RLock()
+	defer p.errLk.RUnlock()
+	if p.err != nil {
+		return p.err
+	}
+	return p.ctx.Err()
+}
+
+func (p *provider) worker() {
+	for {
+		select {
+		case id := <-p.jobs:
+			err := p.croute.Provide(p.ctx, id, true)
+			if err != nil && err != kbucket.ErrLookupFailure { // Check for error to decrease test log spamming
+				if p.Err() == nil {
+					p.errLk.Lock()
+					p.err = err
+					p.errLk.Unlock()
+				}
+
+				p.log.Error("failed to provide to DHT", "err", err.Error())
+			}
+
+			p.provided()
+		case <-p.ctx.Done():
+			for {
+				select {
+				case <-p.jobs: // drain chan
+					p.provided() // ensure done is closed
+				default:
+					return
+				}
+			}
+		case <-p.done:
+			return
+		}
+	}
+}
+
+func (p *provider) provided() {
+	if atomic.AddInt32(&p.total, -1) == 0 {
```

Comment:
It may look unfamiliar if you are not comfortable with the atomic package. I tend to think about reducing thread/goroutine blocking when I write async code, so I use atomic where possible. Furthermore, for this case using atomics makes code smaller.

Reply:
I'm missing something here, how does it block less or make the code smaller? 

While it is my opinion that we should use the standard way of doing things when possible, especially if it's a quick fix, whether we utilize the logic of a channel or re implement it here is arbitrary, and I'm leaving it up to you.

Reply:
> so I use atomic where possible.

Quoting the atomic package documentation:
> These functions require great care to be used correctly. Except for special, low-level applications, synchronization is better done with channels or the facilities of the sync package. Share memory by communicating; don't communicate by sharing memory.

https://golang.org/pkg/sync/atomic/

I feel like the code should contain a very clear argument (e.g. as a comment in the code) why this was preferred over channels. Not only would that help answer @evan-forbes question but it would also would the code speak for itself.

> It may look unfamiliar if you are not comfortable with the atomic package.

The atomic package will be unfamiliar to most go engineers as they likely won't need it during their career.

Reply:
> (Don't think mentioning here is a good idea).

It's just part of the atomic package docs.

>  I will rewrite this to use channels in singleton Provider or maybe consider using some simple 3rd-party lib for worker pool, like https://github.com/gammazero/workerpool, as it matches ideally for our use-case.

If we can do this simply without introducing another dependency that is preferred.

> I don't see any reason to rewrite this now, as this is not the final code and it works well.

I think it wasn't always clear which parts are temporary and will be changed asap. Note that there is always the risk that we just forget about this if not done in the PR, or immediately afterwards.